File: MLanguage/M/lexer.py
from enum import Enum

# Let x = 45 + (foo * bar)

class TokenType(Enum):
    Number = 0
    Identifier = 1
    Equals = 2
    OpenParen = 3
    CloseParen = 4
    BinaryOperator = 5
    Let = 6

KEYWORDS = {
    "let": TokenType.Let,
}

class Token:
    def __init__(self, value: str, type_: TokenType):
        self.value = value
        self.type = type_

    def __repr__(self):
        return f"Token(value='{self.value}', type={self.type.name})"

def token(value="", type_=TokenType.Identifier):
    return Token(value, type_)

def isalpha(char: str):
    return char.isalpha()

def isint(char: str):
    return char.isdigit()

def isskippable(char: str):
    return char in " \n\t"

def tokenize(source_code: str):
    tokens = []
    src = list(source_code)

    while src:
        current = src[0]
        if current == '(':
            tokens.append(token(src.pop(0), TokenType.OpenParen))
        elif current == ')':
            tokens.append(token(src.pop(0), TokenType.CloseParen))
        elif current in "+-*/":
            tokens.append(token(src.pop(0), TokenType.BinaryOperator))
        elif current == "=":
            tokens.append(token(src.pop(0), TokenType.Equals))
        elif isint(current):
            num = ''
            while src and isint(src[0]):
                num += src.pop(0)
            tokens.append(token(num, TokenType.Number))
        elif isalpha(current):
            ident = ''
            while src and isalpha(src[0]):
                ident += src.pop(0)
            reserved = KEYWORDS.get(ident)
            if reserved is None:
                tokens.append(token(ident, TokenType.Identifier))
            else:
                tokens.append(token(ident, reserved))
        elif isskippable(current):
            src.pop(0)
        else:
            print("Unrecognized char found in source:", current)
            exit()

    return tokens

# Example usage
if __name__ == "__main__":
    with open("M/test.txt", "r") as f:
        source = f.read()
    for t in tokenize(source):
        print(t)

File: MLanguage/MLanguage/frontend/lexer.ts
export enum TokenType {
  Number,
  string,
  Identifier,

  Let,
  Const,
  Function,
  If,
  While,
  Else,

  BinaryOperator,

  GreaterThan,
  LessThan,
  GreaterEqual,
  LessEqual,
  DoubleEqual,
  NotEquals,
  Equals,

  Comma,
  Dot,
  Colon,
  Semicolon,
  OpenParen,
  CloseParen,
  OpenBrace,
  OpenBracket,
  CloseBrace,
  CloseBracket,

  EOF,
}

const KEYWORDS: Record<string, TokenType> = {
  let: TokenType.Let,
  const: TokenType.Const,
  function: TokenType.Function,
  if: TokenType.If,
  while: TokenType.While,
  else: TokenType.Else,
};

export interface Token {
  value: string;
  type: TokenType;
}

function token(value = "", type: TokenType): Token {
  return { value, type };
}

function isalpha(src: string) {
  return src.toUpperCase() != src.toLowerCase();
}

function isskippable(str: string) {
  return str == " " || str == "\n" || str == "\t" || str == "\r";
}

function isint(str: string) {
  const c = str.charCodeAt(0);
  const bounds = ["0".charCodeAt(0), "9".charCodeAt(0)];
  return c >= bounds[0] && c <= bounds[1];
}

export function tokenize(sourceCode: string): Token[] {
  const tokens = new Array<Token>();
  const src = sourceCode.split("");

  while (src.length > 0) {
    // --- Parentheses, braces, brackets ---
    if (src[0] == "(") {
      tokens.push(token(src.shift(), TokenType.OpenParen));
    } else if (src[0] == ")") {
      tokens.push(token(src.shift(), TokenType.CloseParen));
    } else if (src[0] == "{") {
      tokens.push(token(src.shift(), TokenType.OpenBrace));
    } else if (src[0] == "}") {
      tokens.push(token(src.shift(), TokenType.CloseBrace));
    } else if (src[0] == "[") {
      tokens.push(token(src.shift(), TokenType.OpenBracket));
    } else if (src[0] == "]") {
      tokens.push(token(src.shift(), TokenType.CloseBracket));
    }

    // --- Strings ---
    else if (src[0] === '"' || src[0] === "'") {
      const quote = src.shift();
      let str = "";
      let escaped = false;

      while (src.length > 0) {
        const ch = src.shift();
        if (escaped) {
          if (ch === "n") str += "\n";
          else if (ch === "r") str += "\r";
          else if (ch === "t") str += "\t";
          else str += ch;
          escaped = false;
          continue;
        }

        if (ch === "\\") {
          escaped = true;
          continue;
        }

        if (ch === quote) break;
        str += ch;
      }

      tokens.push(token(str, TokenType.string));
    }

    // --- Arithmetic operators ---
    else if (
      src[0] == "+" ||
      src[0] == "-" ||
      src[0] == "*" ||
      src[0] == "/" ||
      src[0] == "%"
    ) {
      tokens.push(token(src.shift(), TokenType.BinaryOperator));
    }

    // --- Comparison and equality operators (multi-char first!) ---
    else if (src[0] == ">" && src[1] == "=") {
      src.shift();
      src.shift();
      tokens.push(token(">=", TokenType.GreaterEqual));
    } else if (src[0] == "<" && src[1] == "=") {
      src.shift();
      src.shift();
      tokens.push(token("<=", TokenType.LessEqual));
    } else if (src[0] == "=" && src[1] == "=") {
      src.shift();
      src.shift();
      tokens.push(token("==", TokenType.DoubleEqual));
    } else if (src[0] == "!" && src[1] == "=") {
      src.shift();
      src.shift();
      tokens.push(token("!=", TokenType.NotEquals));
    } else if (src[0] == ">") {
      tokens.push(token(src.shift(), TokenType.GreaterThan));
    } else if (src[0] == "<") {
      tokens.push(token(src.shift(), TokenType.LessThan));
    } else if (src[0] == "=") {
      tokens.push(token(src.shift(), TokenType.Equals)); // assignment
    }

    // --- Punctuation ---
    else if (src[0] == ";") {
      tokens.push(token(src.shift(), TokenType.Semicolon));
    } else if (src[0] == ":") {
      tokens.push(token(src.shift(), TokenType.Colon));
    } else if (src[0] == ",") {
      tokens.push(token(src.shift(), TokenType.Comma));
    } else if (src[0] == ".") {
      tokens.push(token(src.shift(), TokenType.Dot));
    }

    // --- Numbers ---
    else if (isint(src[0])) {
      let num = "";
      while (src.length > 0 && isint(src[0])) {
        num += src.shift();
      }
      tokens.push(token(num, TokenType.Number));
    }

    // --- Identifiers & keywords ---
    else if (isalpha(src[0])) {
      let ident = "";
      while (src.length > 0 && isalpha(src[0])) {
        ident += src.shift();
      }

      const reserved = KEYWORDS[ident];
      if (typeof reserved == "number") {
        tokens.push(token(ident, reserved));
      } else {
        tokens.push(token(ident, TokenType.Identifier));
      }
    }

    // --- Whitespace ---
    else if (isskippable(src[0])) {
      src.shift();
    }

    // --- Unknown characters ---
    else {
      console.error(
        "Unrecognized character in source:",
        src[0].charCodeAt(0),
        src[0]
      );
      Deno.exit(1);
    }
  }

  tokens.push({ type: TokenType.EOF, value: "EndOfFile" });
  return tokens;
}
